vumi/transports/turn/turn.py

```diff
@@ -0,0 +1,205 @@
+                    transport_metadata={
+                        'original_message_id': message_id,
+                        'timestamp': iso8601(message["timestamp"]),
+                        'network_id': request.args['provider'][0],
+                        'keyword': request.args['keyword'][0],
+                        },
+                    to_addr=normalize_msisdn(self.config['to_addr']),
+                    from_addr=normalize_msisdn(message['from']),
+                    content=content,
+                    )
+                log.msg("Inbound Enqueued.")
+
+            for event in request.args.get("statuses", []):
+                message_id = '%s.%s' % (self.transport_name, event['id'])
+
+                if event["status"] == 'sent':
+                    delivery_status = 'pending'
+                elif event["status"] == 'failed':
+                    delivery_status = 'failed'
+                elif status == 'delivered':
+                    delivery_status = 'delivered'
+                else:
+                    continue
+
+                yield self.publish_event_func(
+                    user_message_id=message_id,
+                    delivery_status=delivery_status,
+                    transport_metadata={
+                        'delivery_status': event["status"],
+                        'timestamp': iso8601(event['timestamp']),
+                        },
+                    to_addr=normalize_msisdn(event['recipient_id']),
+                    )
+        except KeyError, e:
+            request.setResponseCode(http.BAD_REQUEST)
+            msg = ("Need more request keys to complete this request. \n\n"
+                   "Missing request key: %s" % (e,))
+            log.msg('Returning %s: %s' % (http.BAD_REQUEST, msg))
+            request.write(msg)
+        except ValueError, e:
+            request.setResponseCode(http.BAD_REQUEST)
+            msg = "ValueError: %s" % e
+            log.msg('Returning %s: %s' % (http.BAD_REQUEST, msg))
+            request.write(msg)
+        except Exception, e:
+            request.setResponseCode(http.INTERNAL_SERVER_ERROR)
+            log.err("Error processing request: %s" % (request,))
+        request.finish()
+
+    def render(self, request):
+        self.do_render(request)
+        return NOT_DONE_YET
+
+
+class HttpResponseHandler(Protocol):
+    def __init__(self, deferred):
+        self.deferred = deferred
+        self.stringio = StringIO()
+
+    def dataReceived(self, bytes):
+        self.stringio.write(bytes)
+
+    def connectionLost(self, reason):
+        self.deferred.callback(self.stringio.getvalue())
+
+
+class TurnTransport(Transport):
+
+    agent_factory = None  # For swapping out the Agent we use in tests.
+
+    def mkres(self, cls):
+        resource = cls(self.config, self.publish_delivery_report, self.publish_message)
+        self._resources.append(resource)
+        return (resource, self.config['web_%s_path' % (path_key,)])
```

Comment:
If we use the `HttpRpcTransport` as a base for this (even though it's not RPC), then we don't need to handle all of this ourselves. You can look at https://github.com/praekeltfoundation/vumi-vas2nets/blob/develop/vxvas2nets/vas2nets_sms.py for a good example of this.